server/src/main/java/io/spine/server/aggregate/Aggregate.java

```diff
@@ -448,98 +459,11 @@ Snapshot makeSnapshot() {
-            builder.add(record);
-
-            selectedCount++;
-            if (selectedCount == batchSize) {
-                return builder.build();
-            }
-        }
-
-        return builder.build();
-    }
-
-    /**
-     * Creates a new list mapping every {@link Event} in the provided collection to a newly created 
-     * {@link AggregateEventRecord}.
-     */
-    private static List<AggregateEventRecord> toAggregateEventRecords(Collection<Event> events) {
-        final ImmutableList.Builder<AggregateEventRecord> items = ImmutableList.builder();
-        for (Event event : events) {
-            final AggregateEventRecord record = newAggregateEventRecord(event);
-            items.add(record);
-        }
-        return items.build();
-    }
-
-    /**
-     * Creates a new {@link AggregateEventRecord} holding the specified {@link Event}.
-     */
-    private static AggregateEventRecord newAggregateEventRecord(Event event) {
-        final EventContext context = event.getContext();
-        return AggregateEventRecord.newBuilder()
-                                   .setTimestamp(context.getTimestamp())
-                                   .setEvent(event)
-                                   .build();
-    }
-
-    /**
-     * Checks if the command was already handled by the aggregate since last snapshot.
-     *
-     * <p> The check is performed by searching for an event caused by this command that was
-     * committed since last snapshot.
-     *
-     * <p> This functionality support the ability to stop duplicate commands from being dispatched
-     * to the aggregate.
-     *
-     * @param command the command to check
-     * @return {@code true} if the command was handled since last snapshot, {@code false} otherwise
-     */
-    public boolean didHandleSinceLastSnapshot(CommandEnvelope command) {
-        final CommandId newCommandId = command.getId();
-        final AggregateRecordQueryCriteria query = new AggregateRecordQueryCriteria(MAX_VALUE);
-        final Iterator<AggregateEventRecord> iterator = historyBackward(query);
-        while (iterator.hasNext()) {
-            final AggregateEventRecord record = iterator.next();
-            switch (record.getKindCase()) {
-                case EVENT:
-                    final Event event = record.getEvent();
-                    final CommandId eventRootCommandId = getRootCommandId(event);
-                    if (newCommandId.equals(eventRootCommandId)) {
-                        return true;
-                    }
-                    break;
-                case SNAPSHOT:
-                    return false;
-                case KIND_NOT_SET:
-                default:
-                    throw newIllegalStateException("Aggregate record with kind not set.");
-            }
-        }
-        return false;
+    protected Iterator<Event> historyBackward() {
+        final ImmutableList<Event> events = ImmutableList.copyOf(historySinceLastSnapshot);
```

Comment:
Please encapsulate the impl. details into `RecentHistory` class, as we discussed vocally.